controllers/gatekeepersync/gatekeeper_constraint_sync.go

```diff
@@ -0,0 +1,426 @@
+// Copyright Contributors to the Open Cluster Management project
+
+package gatekeepersync
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	depclient "github.com/stolostron/kubernetes-dependency-watches/client"
+	admissionregistration "k8s.io/api/admissionregistration/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	policyv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"open-cluster-management.io/governance-policy-framework-addon/controllers/utils"
+)
+
+const (
+	ControllerName        = "gatekeeper-constraint-status-sync"
+	GatekeeperWebhookName = "gatekeeper-validating-webhook-configuration"
+)
+
+var log = logf.Log.WithName(ControllerName)
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GatekeeperConstraintReconciler) SetupWithManager(mgr ctrl.Manager, constraintEvents *source.Channel) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&policyv1.Policy{}).
+		WithEventFilter(policyPredicates()).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 5}).
```

Comment:
There isn't concurrency in the other controllers that I can see. Concurrency seems good, but I'm wondering about the effect here.

Reply:
@dhaiducek you're right. I added concurrency here because by default, the Gatekeeper audits run every minute so if the user has a lot of constraints, I didn't want there to be a delay in getting results.

The way the concurrency works in controller-runtime is that there is only one goroutine processing a particular reconcile request at a time. For example, since we are reconciling `Policy` objects, if there are three reconciles requests for `Policy` named `a` and one reconcile request for `Policy` named `b`, there would just be one goroutine processing `Policy` `a` and one goroutine processing `Policy` `b`.

In other words, it's concurrent across reconciles for different `Policy` objects but reconciles for the same `Policy` will not run concurrently.